DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/device_global/src/device_global.cpp

```diff
@@ -11,44 +11,41 @@
 
 #include "exception_handler.hpp"
 
-constexpr size_t kNumCounters = 4;
-constexpr int kInitialValue = 10;
-constexpr unsigned kNumIncrements = 3;
+constexpr size_t kNumIterations = 4;
+constexpr unsigned kNumWeightIncrements = 3;
+constexpr unsigned kVectorSize = 4;
 
 namespace exp = sycl::ext::oneapi::experimental;
 
-using IntScalar = std::array<int, 1>;
+using IntScalar = std::array<int, kVectorSize>;
 using FPGAProperties =
-    decltype(exp::properties(exp::device_image_scope, exp::host_access_none));
+    decltype(exp::properties(exp::device_image_scope, exp::host_access_write));
 
-// Array of counters that have a lifetime longer than a single kernel invocation
-exp::device_global<int[kNumCounters], FPGAProperties> counters;
-// Flag if the counters have been initialized - zero-initialized to `false`
-exp::device_global<bool, FPGAProperties> is_counters_initialized;
+// globally declared weights for the calculation
+exp::device_global<int[kVectorSize], FPGAProperties> weights;
 
 // Forward declare the kernel name in the global scope.
 // This FPGA best practice reduces name mangling in the optimization reports.
 class Kernel;
 
-// Launch a kernel that increments the value of a global variable counter
-// at a particular index, and returns the current value of that counter
-void IncrementAndRead(sycl::queue q, IntScalar &result, int index) {
-  sycl::buffer<int, 1> buffer_result(result.data(), 1);
+// Launch a kernel that does a vector weighted add
+// result = a + (weights * b)
+void WeightedVectorAdd(sycl::queue q, IntScalar &a, IntScalar &b, IntScalar &result) {
+  sycl::range<1> io_range(kVectorSize);
+  sycl::buffer buffer_result {result.data(), io_range};
+  sycl::buffer buffer_a{a.data(), io_range};
+  sycl::buffer buffer_b{b.data(), io_range};
```

Comment:
we usually recommend that IP authoring kernels use USM, since it is easier to add `mm_host` customizations to USM pointers than to accessors.